### DISCUSSION TITLE
example.predicted.ents is empty!

### DISCUSSION THREAD

**Discussion by amirj** [explosion/spaCy#8010] [Help: Coding & Implementations] [answered]
I created a dataset containing a list of `spacy.training.Example`:
```python
examples = []
for text, annotation in dataset:
    try:
        doc = nlp.make_doc(text)
        example = Example.from_dict(doc, annotation)
        example.reference = self.sentencizer(example.reference)
        examples.append(example)
```
 I just noticed that `example.predicted.ents` is empty though `example.reference.ents` represents the gold entities/links. Although I can assign a `doc` object to `example.predicted`, but I've been thinking if there is a better way to do that? e.g. when calling `nlp.evaluate(examples)`, the predicted field should be assigned?

**Comment by svlandeg** [comment]
I'm not sure I understand your question. The `predicted` `Doc` object is made with `nlp.make_doc` (it's the first argument to `Example.from_dict`) and so it only contains tokenization. 

When the examples are being evaluated, the `nlp` pipeline [will run](https://github.com/explosion/spaCy/blob/a71194362f4a1f40e158ae216f9e3f852347f53e/spacy/language.py#L1352) on the `predicted` object and the predictions will be stored on that `Doc`. If `example.predicted.ents` is still empty after you've run your NLP pipeline on it, can you double check that you have an NER in the pipeline, and that it's actually properly trained & predicting things?

**Comment by adrianeboyd** [comment, marked as answer]
Do you mean providing the ents for evaluation for the entity linker? You currently have to do it by hand when constructing the examples. I think depending on your task/evaluation you can choose whether you'd like to evaluate with gold or predicted ents (Sofie may know more details here):

```
# gold ents
example.predicted.ents = example.reference.ents
# predicted ents, create predicted doc for example using `nlp` containing `ner`
doc = nlp(text)
```

One confusing thing is that `nlp.evaluate` runs on an internal copy of the predicted doc, so you can't see the predictions afterwards. If you want to see the predictions, you need to use `nlp.pipe`, so something like this:

```python
docs = list(nlp.pipe([ex.reference.text for ex in examples]))
```

In the upcoming v3.1, you'll be able to set up the config so that an existing NER component is applied during training to get predicted ents without any extra steps when constructing examples. Then you can also use the default `spacy.Corpus.v1` reader instead of constructing examples in a custom reader or custom training loop.
